R/02-plotTitles.R

```diff
@@ -0,0 +1,223 @@
+    # if null: take values from config
+    initTitles <- list(
+      plot = defaultTitleFormat(type = type),
+      xAxis = defaultTitleFormat(type = type),
+      yAxis = defaultTitleFormat(type = type)
+    )
+  }
+
+  ns <- NS(id)
+  tagList(
+    h4("Titles"),
+    selectInput(
+      inputId = ns("labelName"),
+      label = "Label",
+      choices = c(
+        "plot title" = "plot",
+        "x axis" = "xAxis",
+        "y axis" = "yAxis"
+      ),
+      selected = "plot"
+    ),
+    textInput(ns("text"), label = "Text",
+              value = initTitles[["plot"]][["text"]],
+              placeholder = "Custom title ..."),
+    colourInput(ns("color"),
+                label = "Text color",
+                value = initTitles[["plot"]][["color"]]),
+    selectInput(
+      ns("fontType"),
+      label = "Font type",
+      choices = fontChoicesSelect(type = type),
+      selected = initTitles[["plot"]][["fontType"]]),
+    sliderInput(
+      ns("size"),
+      label = "Text size",
+      value = initTitles[["plot"]][["size"]],
+      min = sizeValuesSlider(type = type)[["min"]],
+      max = sizeValuesSlider(type = type)[["max"]],
+      step = sizeValuesSlider(type = type)[["step"]]
+    ),
+    checkboxInput(
+      inputId = ns("hide"),
+      label = "Hide label",
+      value = initTitles[["plot"]][["hide"]],
+      width = "100%"
+    )
+  )
+}
+
+#' Server function for plot titles
+#'
+#' Backend for plot titles module
+#'
+#' @param id namespace id
+#' @param type (character) Type of the plot to add titles to, one of "none", "ggplot", "base".
+#' @inheritParams plotExportServer
+#'
+#' @export
+plotTitlesServer <- function(id, type = c("none", "ggplot", "base"), initTitles = NULL) {
+  type <- match.arg(type)
+
+  moduleServer(id,
+               function(input, output, session) {
+                 if (is.null(initTitles)) {
+                   # if null: take values from config
+                   titles <- reactiveValues(
+                     plot = defaultTitleFormat(type = type),
+                     xAxis = defaultTitleFormat(type = type),
+                     yAxis = defaultTitleFormat(type = type)
+                   )
+                 } else if (inherits(initTitles, "list")) {
+                   titles <- reactiveValues(
+                     plot = initTitles[["plot"]],
+                     xAxis = initTitles[["xAxis"]],
+                     yAxis = initTitles[["yAxis"]]
+                   )
+                 } else {
+                   titles <- initTitles
+                 }
+
+                 if (type == "none") return(titles)
+
+                 observe({
+                   req(input[["labelName"]])
+                   updateUserInputs(id, input = input, output = output, session = session,
+                                    userInputs = titles[[input[["labelName"]]]])
+                 }) %>%
+                   bindEvent(input[["labelName"]])
+
+                 observe({
```

Comment:
Just an idea because most functions are quite long. We could move all of the observer from line 103 to 131 into a wrapper function (with a meaningful name). We could then only call this function here which would improve readability a lot. Maybe this can also be done for other modules.